Documentation of l3afd user configuration options (#84)
l3af-project/l3afd: docs/configdoc.md

# L3AFD Config Options Documentation

See [l3afd.cfg](https://github.com/l3af-project/l3afd/blob/main/config/l3afd.cfg) for a full example configuration.

```
[DEFAULT]

[l3afd]
pid-file: ./l3afd.pid
datacenter: dummy
bpf-dir: /dev/shm
bpf-log-dir:
kernel-major-version: 4
kernel-minor-version: 15
shutdown-timeout: 1s
http-client-timeout: 10s
max-nf-restart-count: 3
bpf-chaining-enabled: true
swagger-api-enabled: false
# PROD | DEV
environment: PROD
....
```

### Below is the detailed documentation for each field

## [l3afd]

> | FieldName     | Example       | Description     |

Reviewer: For each field below, state whether it is required or optional.  If the answer varies by field, then add a column to this table.

| ------------- | ------------- | --------------- |

> |pid-file| `"l3afd.pid"`  |It contains process id of eBPF Programs |

Reviewer: I'm guessing you mean
```suggestion
|pid-file| `"l3afd.pid"`  |The name of a file that contains process id of eBPF Programs |
```
But is a path legal or only the filename?   Where is the format of this file documented?
Is the file something that l3afd writes to?  (I assume so, but say so)

> |datacenter| `"dummy"` | Name of Datacenter|

Reviewer: What is this used for?  Explain.

|bpf-dir| `"/dev/shm"` | Absolute Path of eBPF Programs (where they are extracted) |

> |bpf-log-dir|`""`      | Absolute Path,where we store logs eBPF Programs|

Reviewer: ```suggestion
|bpf-log-dir|`""`      | Absolute Path, where we store logs of eBPF Programs|
```
Clarify... is this a log of l3afd's actions regarding ebpf programs?

> |kernel-major-version|`"4"`|Major version of the kernel|

Reviewer: Is this used on Windows too?  Or is it only for Linux?

> |kernel-minor-version|`"15"`|Miner version of the kernel (Ex 4.15)|

Reviewer: ```suggestion
|kernel-minor-version|`"15"`|Minor version of the kernel (Ex 4.15)|
```

> |shutdown-timeout|`"1s"`|It is the amount of time allowed to stop all network functions|

Reviewer: ```suggestion
|shutdown-timeout|`"1s"`|The amount of time allowed to stop all eBPF programs|
```
What does "allowed" mean?   What happens after that time elapses?

> |http-client-timeout|`"10s"`|It is the amount of time allowed to get response headers|

Reviewer: To get response headers from whom?  Is this about the REST API exposed by l3afd?  Or about the ability to fetch a program from a repository via HTTP?

> |max-nf-restart-count|`"3"`|Maximum number of attempts to monitor eBPF Program|

Reviewer: "nf" is a misnomer so should be renamed.
But what is an "attempt to monitor eBPF program"?  Explain.

> |bpf-chaining-enabled|`"true"`|Boolean to set bpf-chaining |

Reviewer: Explain (or reference a document that does).  What does it mean when this is true and what does it mean when this is false?  I.e., what will l3afd do?

> |swagger-api-enabled|`"false"`|Boolean to check swagger api enable or not|

Reviewer: ```suggestion
|swagger-api-enabled|`"false"`|Boolean to check whether the swagger API is enabled or not|
```
Add link to more info?   Is this about whether l3afd exposes a REST API via http?
How do you interact with l3afd if this is false?

> |environment|`"PROD"`| Environment  variable |

Reviewer: environment variable for what?

> |BpfMapDefaultPath|`"/sys/fs/bpf"`|It is a base path for storing bpf maps|

Reviewer: ```suggestion
|BpfMapDefaultPath|`"/sys/fs/bpf"`|The base pin path for eBPF maps|
```

> |TcMapsRelativePath|`"tc/globals"`|It is a relative path of tc maps with respect to BpfMapDefaultPath|

Reviewer: ```suggestion
|TcMapsRelativePath|`"tc/globals"`|The relative pin path of tc maps with respect to BpfMapDefaultPath|
```
Also, why does tc have a relative path but xdp does not?  Is this parameter required or optional?

Author: Basically xdp maps are stored in BpfMapDefaultPath , hence there is no relative path for xdp programs.

Reviewer: Why can't tc maps be stored in BpfMapDefaultPath then just like for xdp?

## [kf-repo]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |

> |url| `"http://localhost:8000/"`|Default Repository to Download eBPF Programs|

Reviewer: ```suggestion
|url| `"http://localhost:8000/"`|Default repository from which to download eBPF programs|
```

## [web]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |
|metrics-addr|`"0.0.0.0:8898"`|Prometheus endpoint for pull/scrape the metrics|

> |kf-poll-interval|`"30s"`|Periodic Interval to scrape metrics using Prometheus|

Reviewer: Add link to prometheus documentation
Also "kf" in kf-poll-interval needs to be renamed (in a separate PR)

|n-metric-samples|`"20"`|Number of Metric Samples|

## [admind]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |

> |host|`""`|Hostname of admin|

Reviewer: What does this mean?

|username|`""`|Username of admin|

> |api-key|`""`|Api key of admin|

Reviewer: API key for what API?  How is this used?

> |groud-id|`"0"`|Group id of admin|

Reviewer: What is this used for?  Does it apply to Windows too and if so what format should group-id be in?

> |api-enabled|`"true"`|Boolean to check api enabled or not|

Reviewer: Which API?  How is this different from swagger-api-enabled?

## [xdp-root-program]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |
|name|`"xdp-root"`|Name of xdp-root program|

> |artifact|`"xdp-root.tar.gz"`|Archive of xdp-root program|

Reviewer: What formats are legal here?

> |ingress-map-name|`"root_array"`|Ingress map name of xdp-root program|

Reviewer: What is the "ingress map"?  Maybe add a link to documentation?

|command|`"xdp_root"`|Command to run xdp-root program|

> |version|`"1.01"`|Version of xdp-root program|

Reviewer: What does l3afd use the version for?

Author: When we are downloading  programs the artifact path is something  like  (Repourl+ProgramName + ProgramVersion + Platform + Artifact name) this.
we are just storing programs with there versions in the repository that's why  we need version.

> |is-user-program|`"false"`|Boolean to check xdp-root is kernel space or user space program|

Reviewer: Line 73 has a command to run xdp-root, so what does it mean for this to be true?  Can you give an example of what command would contain in that case?

Author: In my knowledge  for 
Linux :  sudo xdp_root
Windows : runas.exe /user:administrator xdp_root

## [tc-root-program]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |
|name|`"tc_root"`|Name of tc_root program|

> |artifact|`"l3af_tc_root.tar.gz"`|Archive of tc_root program|

Reviewer: Lines 70-71 used hyphens in xdp-root.  Can we be consistent instead of using underscores here?

Author: We need to do series of changes for getting this consistency .
I think we can do this task in seperate PR

|ingress-map-name|`"tc_ingress_root_array"`|Ingress map name of tc_root program|

> |egress-map-name|`"tc_egress_root_array"`|Egress map name of tc_root program|

Reviewer: What is an "egress map"?   Maybe provide a link to other documentation about the tc-root program requirements?

|command|`"tc_root"`|Command to run tc_root program|
|version|`"1.0"`|Version of tc_root program|

> |is-user-program|`"false"`|Boolean to check tc_root is kernel space or user space program|

Reviewer: Same question as on line 75

Author: check comments on previous suggestion

## [ebpf-chain-debug]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |

> |addr|`"0.0.0.0:8899"`|Address of ebpf chain debuging|
> |enabled|`"true"`|Boolean to check ebpf-chain-debuging is enabled or not|

Reviewer: ```suggestion
|addr|`"0.0.0.0:8899"`|Address of ebpf chain debugging|
|enabled|`"true"`|Boolean to check ebpf-chain-debugging is enabled or not|
```

What is "chain debugging"?
If the address is IPv6, what format is it in?  I.e., are square braces used like in "[::]:8899"?   Are IPv6 scoped addresses legal or illegal?
Can a hostname be used or must it be an address literal?

## [l3af-configs]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |

> |restapi-addr|`"localhost:53000"`| Address of l3af-configs rest api |

Reviewer: "localhost:53000" is not an address, it's a hostname and a port.
Is the port required?  Is an address legal instead of a hostname?
If so, when IPv6 is used, are square braces used?

# [l3af-config-store]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |

> |filename|`"/etc/l3afd/l3af-config.json"`|Absolute path of persistent config file|

Reviewer: What is the format of l3af-config.json contents?

# [mtls]
| FieldName     | Example       | Description     |
| ------------- | ------------- | --------------- |

> |enabled| `"true"` | Boolean to check mtls enabled or not|

Reviewer: Is mtls used for both the REST API exposed by l3afd as well as for downloading programs from a repository, or only one of those?

> |min-tls-version|`""`| Minimum tls version allowed|

Reviewer: What does it mean for this to be empty?  Shouldn't this say
```suggestion
|min-tls-version|`"1.3"`| Minimum tls version allowed|
```
?

> |cert-dir|`"/etc/l3af/certs"`|Absolute path of ca certificates |

Reviewer: Presumably certificates are always stored as files in the filesystem, yes?   Note that is not best practice on Windows (see https://docs.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores) so is worth stating explicitly if this is a limitation of l3afd.

|server-crt-filename|`"server.crt"`|Server's ca certificate filename|

> |server-key-filename|`"server.key"`|Server's key filename|

Reviewer: "public" key?
